dev/release/verify_rc.sh

```diff
@@ -0,0 +1,227 @@
+#!/usr/bin/env bash
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+set -eu
+
+SOURCE_DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+TOP_SOURCE_DIR="$(dirname "$(dirname "${SOURCE_DIR}")")"
+
+if [ "$#" -ne 2 ]; then
+  echo "Usage: $0 <version> <rc>"
+  echo " e.g.: $0 18.0.0 1"
+  exit 1
+fi
+
+set -o pipefail
+set -x
+
+VERSION="$1"
+RC="$2"
+
+ARROW_DIST_BASE_URL="https://dist.apache.org/repos/dist/dev/arrow"
+DOWNLOAD_RC_BASE_URL="https://github.com/apache/arrow-go/releases/download/v${VERSION}-rc${RC}"
+ARCHIVE_BASE_NAME="apache-arrow-go-${VERSION}"
+
+: "${VERIFY_DEFAULT:=1}"
+: "${VERIFY_DOWNLOAD:=${VERIFY_DEFAULT}}"
+: "${VERIFY_FORCE_USE_GO_BINARY:=0}"
+: "${VERIFY_SIGN:=${VERIFY_DEFAULT}}"
+
+VERIFY_SUCCESS=no
+
+setup_tmpdir() {
+  cleanup() {
+    go clean -modcache || :
+    if [ "${VERIFY_SUCCESS}" = "yes" ]; then
+      rm -rf "${VERIFY_TMPDIR}"
+    else
+      echo "Failed to verify release candidate. See ${VERIFY_TMPDIR} for details."
+    fi
```

Comment:
instead of calling `go clean -modcache` we could try just setting the `GOCACHE` and `GOMODCACHE` env vars to point to directories in the tmpdir. That way we still get clean caches, but we don't blow up the users cache dirs.

Reply:
Removing cache isn't the main purpose of it.
It's for avoiding a permission problem. See also: https://github.com/apache/arrow/pull/38222

BTW, do we need to set `GOCACHE`/`GOMODCACHE` explicitly when we set `GOPATH`? Can `GOCACHE`/`GOMODCACHE` be set automatically under `GOPATH`?

Reply:
Ah gotcha. 

> BTW, do we need to set `GOCACHE`/`GOMODCACHE` explicitly when we set `GOPATH`? Can `GOCACHE`/`GOMODCACHE` be set automatically under `GOPATH`?

`GOMODCACHE` should generally be `${GOPATH}/pkg/mod`, we might not need to change it explicitly since we set `GOPATH`.

`GOCACHE` can be anywhere though and typically is set to `~/.cache/go-build`, if we're not running `go clean -cache` or otherwise cleaning the build cache, and only using `go clean -modcache` then we're probably fine here without issue.

Reply:
OK. Let's use `go clean -modcache` here.